path_oidc.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return nil, err
+	}
+
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return logical.ErrorResponse("missing redirect_uri"), nil
+	}
+
+	if role != nil && role.RoleType == "oidc" {
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		if len(role.AllowedRedirectURIs) == 0 ||
+			strutil.StrListContains(role.AllowedRedirectURIs, redirectURI) {
+
+			// "openid" is a required scope for OpenID Connect flows
+			scopes := append([]string{oidc.ScopeOpenID}, role.OIDCScopes...)
+
+			// Configure an OpenID Connect aware OAuth2 client
+			oauth2Config := oauth2.Config{
+				ClientID:     config.OIDCClientID,
+				ClientSecret: config.OIDCClientSecret,
+				RedirectURL:  redirectURI,
+				Endpoint:     provider.Endpoint(),
+				Scopes:       scopes,
+			}
+
+			stateID, nonce, err := b.createState(roleName, redirectURI)
+			if err != nil {
+				return nil, errwrap.Wrapf("error generating OAuth state: {{err}}", err)
+			}
+
+			authCodeURL = oauth2Config.AuthCodeURL(stateID, oidc.Nonce(nonce))
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"auth_url": authCodeURL,
+		},
+	}
+
+	return resp, nil
+}
+
+// createState make an expiring state object, associated with a random state ID
+// that is passed throughout the OAuth process. A nonce is also included in the
+// auth process, and for simplicity will be identical in length/format as the state ID.
+func (b *jwtAuthBackend) createState(rolename, redirectURI string) (string, string, error) {
+	randstr, err := base62.Random(2 * stateLength)
+	if err != nil {
+		return "", "", err
+	}
+
+	stateID, nonce := randstr[0:stateLength], randstr[stateLength:]
+
+	b.l.Lock()
+	b.oidcStates[stateID] = &oidcState{
+		rolename:    rolename,
+		expiration:  time.Now().Add(oidcStateTimeout),
+		nonce:       nonce,
+		redirectURI: redirectURI,
+	}
+	b.l.Unlock()
+
+	return stateID, nonce, nil
+}
+
+// verifyState tests whether the provided state ID is valid and returns the
+// associated state object if so. A nil state is returned if the ID is not found
+// or expired. The state should only ever be retrieved once and is deleted as
+// part of this request.
+func (b *jwtAuthBackend) verifyState(state string) *oidcState {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	s := b.oidcStates[state]
+	if s != nil && time.Now().After(s.expiration) {
+		s = nil
+	}
+
+	delete(b.oidcStates, state)
+
+	return s
+}
+
+// stateGC will start a goroutine which periodically deletes all expired states
+// that were never removed via the normal request process.
+func (b *jwtAuthBackend) stateGC() {
+	go func() {
+		for {
+			now := time.Now()
+			b.l.Lock()
+			for k := range b.oidcStates {
+				if now.After(b.oidcStates[k].expiration) {
+					delete(b.oidcStates, k)
+				}
+			}
+			b.l.Unlock()
+			time.Sleep(1 * time.Minute)
```

Comment:
Since the sleep is in a loop, might be nice to consider using a ticker that goes off every minute. If it were part of a `select`, it would also be possible to listen to some sort of `shutdown` channel or a call context cancellation channel simultaneously, and to return from the goroutine if they were called.

Reply:
I bet the `ctx` [here](https://github.com/hashicorp/vault-plugin-auth-jwt/blob/895c729e7b08bb8e65e8bc401de6ce0b918da32c/backend.go#L18) would be usable for detecting a graceful shutdown of the enclosing environment.

Reply:
Looks like a request context wouldn't be applicable in this case since it's only used at the backend level.

Reply:
To be clearer, I was imagining this more like this:
```
ticker := yadda
for {
    select {
    case <- ticker.C:
        ....
    case <- ctx.Done():
        ....
    }
}
```

Reply:
I'll look into wiring in context.

Reply:
Just pushed a proposed change that's hooked into the Clean function.